pkg/skaffold/jib/jib.go

```diff
@@ -17,64 +17,163 @@ limitations under the License.
 package jib
 
 import (
+	"encoding/json"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"sort"
+	"strings"
+	"time"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 	"github.com/karrick/godirwalk"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-func getDependencies(cmd *exec.Cmd) ([]string, error) {
-	stdout, err := util.RunCmdOut(cmd)
-	if err != nil {
+// filesTemplate contains fields parsed from Jib's JSON output
+type filesTemplate struct {
+	// Build lists paths to build definitions that trigger a call out to Jib to refresh the filesTemplate, as well as a rebuild, upon changing
+	Build []string
+
+	// Inputs lists paths to build dependencies that trigger a rebuild upon changing
+	Inputs []string
+
+	// Ignore lists paths to files that should be ignored when checking for changes to rebuild
+	Ignore []string
+}
+
+// filesLists contains cached build/input dependencies
+type filesLists struct {
+	// FilesTemplate saves the most recent output of the Jib Skaffold files task/goal
+	FilesTemplate filesTemplate
+
+	// BuildFileTimes keeps track of the last modification time of each build file
+	BuildFileTimes map[string]time.Time
+}
+
+// watchedFiles maps from project name to watched files
+var watchedFiles = map[string]filesLists{}
+
+// getDependencies returns a list of files to watch for changes to rebuild
+func getDependencies(cmd *exec.Cmd, projectName string) ([]string, error) {
+	dependencyList := []string{}
+	template := watchedFiles[projectName].FilesTemplate
+	if len(template.Inputs) == 0 && len(template.Build) == 0 {
+		// Make sure build file modification time map is setup
+		if watchedFiles[projectName].BuildFileTimes == nil {
+			watched := watchedFiles[projectName]
+			watched.BuildFileTimes = make(map[string]time.Time)
+			watchedFiles[projectName] = watched
+		}
+
+		// Refresh dependency list if empty
+		if err := refreshDependencyList(cmd, projectName); err != nil {
+			return nil, err
```

Comment:
Either `return nil, refreshDependencyList(..` or wrap the error to provide context